core/freq_params_traffic/traffic_shaping.go

```diff
@@ -0,0 +1,230 @@
+				newQps := int64(0)
+				if toAddTokenNum+restQps > maxCount {
+					newQps = maxCount - acquireCount
+				} else {
+					newQps = toAddTokenNum + restQps - acquireCount
+				}
+				if newQps < 0 {
+					return base.NewTokenResultBlocked(base.BlockTypeFreqParamsFlow, fmt.Sprintf("defaultTrafficShapingController, the new QPS after subbing acquire(%d) is less than 0.", acquireCount))
+				}
+				if atomic.CompareAndSwapInt64(oldQpsPtr, restQps, newQps) {
+					atomic.StoreInt64(lastAddTokenTimePtr, currentTimeInMs)
+					return base.NewTokenResultPass()
+				}
+				runtime.Gosched()
+			}
+		} else {
+			//check whether the rest of token is enough to acquire
+			oldQpsPtr, found := tokenCounter.Get(arg)
+			if found {
+				oldRestToken := atomic.LoadInt64(oldQpsPtr)
+				if oldRestToken-acquireCount >= 0 {
+					//update
+					if atomic.CompareAndSwapInt64(oldQpsPtr, oldRestToken, oldRestToken-acquireCount) {
+						return base.NewTokenResultPass()
+					}
+				} else {
+					return base.NewTokenResultBlocked(base.BlockTypeFreqParamsFlow, fmt.Sprintf("defaultTrafficShapingController, the rest token is not enough, oldRestToken: %d, acquire: %d", oldRestToken, acquireCount))
+				}
+			}
+			runtime.Gosched()
+		}
+	}
+}
+
+func (c *throttlingTrafficShapingController) performChecking(arg interface{}, acquireCount int64) *base.TokenResult {
+	metric := c.metric
+	if metric == nil {
+		return base.NewTokenResultPass()
+	}
+
+	if c.metricType == Concurrency {
+		return c.performCheckingForConcurrencyMetric(arg)
+	} else if c.metricType > QPS {
+		return base.NewTokenResultPass()
+	}
+
+	timeCounter := metric.ruleTimeCounter
+	tokenCounter := metric.ruleTokenCounter
+	if timeCounter == nil || tokenCounter == nil {
+		return base.NewTokenResultPass()
+	}
+
+	// calculate available token
+	tokenCount := int64(c.threshold)
+	val, existed := c.specificItems[arg]
+	if existed {
+		tokenCount = val
+	}
+	if tokenCount <= 0 {
+		return base.NewTokenResultBlocked(base.BlockTypeFreqParamsFlow, "throttlingTrafficShapingController, the setting tokenCount is <= 0")
+	}
+	intervalCostTime := int64(math.Round(float64(acquireCount * c.durationInSec * 1000 / tokenCount)))
+	for {
+		currentTimeInMs := int64(util.CurrentTimeMillis())
+		lastPassTimePtr := timeCounter.AddIfAbsent(arg, &currentTimeInMs)
+		if lastPassTimePtr == nil {
+			// first access arg
+			return base.NewTokenResultPass()
+		}
+		// load the last pass time
+		lastPassTime := atomic.LoadInt64(lastPassTimePtr)
+		// calculate the expected pass time
+		expectedTime := lastPassTime + intervalCostTime
+
+		if expectedTime <= currentTimeInMs || expectedTime-currentTimeInMs < c.maxQueueingTimeMs {
+			if atomic.CompareAndSwapInt64(lastPassTimePtr, lastPassTime, currentTimeInMs) {
+				awaitTime := expectedTime - currentTimeInMs
+				if awaitTime > 0 {
+					atomic.StoreInt64(lastPassTimePtr, expectedTime)
+					time.Sleep(time.Duration(awaitTime) * time.Millisecond)
```

Comment:
Maybe it's better to use `NewTokenResultShouldWait` instead?

Reply:
Okay